File: src/server/backend/runtime/checkers/tool_after/tool_result.py
"""Checker for tool result events (observation injection)."""
from __future__ import annotations

from shared.schemas.context import RuntimeContext
from shared.schemas.events import EventType, RuntimeEvent
from backend.runtime.checkers.base import BaseChecker, CheckResult
from backend.runtime.checkers.common.patterns import find_signals, text_of
from backend.runtime.checkers.registry import register


@register(
    name="tool_result",
    description="Detect secrets and prompt-injection content in tool results.",
)
class ToolResultChecker(BaseChecker):
    event_types = [EventType.TOOL_RESULT]

    def check(
        self,
        event: RuntimeEvent,
        context: RuntimeContext,
        trajectory_window: list[RuntimeEvent] | None = None,
    ) -> CheckResult:
        text = text_of(event.payload.get("result"))
        signals = find_signals(text)
        if "prompt_injection" in signals:
            signals.append("tool_result_injection")
        return CheckResult(risk_signals=sorted(set(signals)))

File: src/client/js/agentguard/plugins/tool_after/tool_result.js
"use strict";

const { BasePlugin, CheckResult } = require("../base");
const { EventType } = require("../../schemas/events");

class ToolResultPlugin extends BasePlugin {
  constructor() {
    super();
    this.event_types = [EventType.TOOL_RESULT];
  }

  check(event) {
    const text = JSON.stringify((event.payload || {}).result || "");
    const signals = [];
    if (/ignore previous instructions|system prompt/i.test(text)) {
      signals.push("prompt_injection");
    }
    return new CheckResult({ risk_signals: signals });
  }
}

module.exports = {
  ToolResultPlugin,
};
